herddb-core/src/main/java/herddb/core/TableManager.java

```diff
@@ -994,40 +996,44 @@ public default void endNewRecordsInTransactionBlock() {
+                            }
+                        } catch (IllegalArgumentException err) {
+                            locksManager.releaseLock(lockHandle);
+                            writes.add(FutureUtils.exception(new StatementExecutionException(err.getMessage(), err)));
+                            return;
                         }
-                    } catch (IllegalArgumentException err) {
+                    }
+                    final long size = DataPage.estimateEntrySize(actual.key, newValue);
+                    if (size > maxLogicalPageSize) {
                         locksManager.releaseLock(lockHandle);
-                        writes.add(FutureUtils.exception(new StatementExecutionException(err.getMessage(), err)));
+                        writes.add(FutureUtils.exception(new RecordTooBigException("New version of record " + actual.key
+                                + " is to big to be update: new size " + size + ", actual size " + DataPage.estimateEntrySize(actual)
+                                + ", max size " + maxLogicalPageSize)));
                         return;
                     }
-                }
-                final long size = DataPage.estimateEntrySize(actual.key, newValue);
-                if (size > maxLogicalPageSize) {
-                    locksManager.releaseLock(lockHandle);
-                    writes.add(FutureUtils.exception(new RecordTooBigException("New version of record " + actual.key
-                            + " is to big to be update: new size " + size + ", actual size " + DataPage.estimateEntrySize(actual)
-                            + ", max size " + maxLogicalPageSize)));
-                    return;
-                }
 
-                LogEntry entry = LogEntryFactory.update(table, actual.key, Bytes.from_array(newValue), transaction);
-                CommitLogResult pos = log.log(entry, entry.transactionId <= 0);
-                writes.add(pos.logSequenceNumber.thenApply(lsn -> new PendingLogEntryWork (entry, pos, lockHandle)));
-                lastKey.value = actual.key;
-                lastValue.value = newValue;
-                updateCount.incrementAndGet();
-            }
-        }, transaction, true, true);
+                    LogEntry entry = LogEntryFactory.update(table, actual.key, Bytes.from_array(newValue), transaction);
+                    CommitLogResult pos = log.log(entry, entry.transactionId <= 0);
+                    writes.add(pos.logSequenceNumber.thenApply(lsn -> new PendingLogEntryWork (entry, pos, lockHandle)));
+                    lastKey.value = actual.key;
+                    lastValue.value = newValue;
+                    updateCount.incrementAndGet();
+                }
+            }, transaction, true, true);
+        } catch (StatementExecutionException validationError) {
+            return FutureUtils.exception(validationError);
```

Comment:
This is not the right point

That "validationError" should be handled in the same "catch" al line 1101 

your new error is throws from
`byte[] newValue = function.computeNewValue(actual, context, tableContext);`

we are already catching validation errors for 
`RecordSerializer.validatePrimaryKey`

you should enclose
byte[] newValue = function.computeNewValue(actual, context, tableContext);

in the same try/catch.
Your "serialization" error should be an IllegalArgumentException, this way we will handle consistently your new impl (not null columns) and the fact that indexed columns cannot be null (this is only because our BRIN index impl does support null keys, but this is another story).

Reply:
@eolivelli done with the changes. All tests pass.